### DISCUSSION TITLE
Unsolicited PSK extension from server

### DISCUSSION THREAD

**Discussion by zphixon** [openssl/openssl#24567] [Q&A] [answered]
Hello, we have a protocol that implements some session resumption at the application layer. This involves a first TLS session with one component of the software which forwards the connection to another component, where a second TLS session is negotiated; i.e. there are close notify messages before the second handshake.

The first handshake messages both include a PSK extension for reasons I cannot identify. After this TLS session is shut down and the underlying connection is forwarded, the new TLS session's ClientHello message doesn't contain the pre_shared_key extension, but the ServerHello message does.

1. Why is the PSK extension present in the first handshake? We don't call SSL_CTX_set_psk_use_session_callback in the server or the client. Is there another API I don't know about?
2. Why is the PSK extension present from the server in the second handshake, despite being unsolicited by the client?

This behavior was observed with 3.0.12. With 1.1.1u, I found the first handshake had the PSK extension, and the second handshake did not.

<details><summary>
Here's the messages from wireshark.
</summary>

```
Transport Layer Security
    TLSv1.3 Record Layer: Handshake Protocol: Client Hello
        Content Type: Handshake (22)
        Version: TLS 1.0 (0x0301)
        Length: 264
        Handshake Protocol: Client Hello
            Handshake Type: Client Hello (1)
            Length: 260
            Version: TLS 1.2 (0x0303)
            Random: ec5d9ed08f409598db399b06c4d9681244b5739ef15ae82c7a9ee6a4fc419587
            Session ID Length: 32
            Session ID: 7cd71dd7bafdd46105491a0809b5fb291794446b3d5aec2ac957b4b6fe2f4ac9
            Cipher Suites Length: 32
            Cipher Suites (16 suites)
            Compression Methods Length: 1
            Compression Methods (1 method)
            Extensions Length: 155
            Extension: ec_point_formats (len=4)
            Extension: supported_groups (len=22)
            Extension: session_ticket (len=0)
            Extension: encrypt_then_mac (len=0)
            Extension: extended_master_secret (len=0)
            Extension: signature_algorithms (len=48)
            Extension: supported_versions (len=5)
            Extension: psk_key_exchange_modes (len=2)
            Extension: key_share (len=38)
            [JA3 Fullstring: 771,4866-4867-4865-157-53-156-47-49200-49196-49199-49195-10-2-59-1-255,11-10-35-22-23-13-43-45-51,29-23-30-25-24-256-257-258-259-260,0-1-2]
            [JA3: a5d9cb5fe930473a3d1a700ffd35f978]
Transport Layer Security
    TLSv1.3 Record Layer: Handshake Protocol: Server Hello
        Content Type: Handshake (22)
        Version: TLS 1.2 (0x0303)
        Length: 128
        Handshake Protocol: Server Hello
            Handshake Type: Server Hello (2)
            Length: 124
            Version: TLS 1.2 (0x0303)
            Random: bab4023402e409520eea833b2fbe57b908ecb9f4466a4f0cd5ac7529669a7af4
            Session ID Length: 32
            Session ID: 7cd71dd7bafdd46105491a0809b5fb291794446b3d5aec2ac957b4b6fe2f4ac9
            Cipher Suite: TLS_AES_256_GCM_SHA384 (0x1302)
            Compression Method: null (0)
            Extensions Length: 52
            Extension: supported_versions (len=2)
            Extension: key_share (len=36)
            Extension: pre_shared_key (len=2)
            [JA3S Fullstring: 771,4866,43-51-41]
            [JA3S: 2253c82f03b621c5144709b393fde2c9]
    TLSv1.3 Record Layer: Change Cipher Spec Protocol: Change Cipher Spec
    TLSv1.3 Record Layer: Handshake Protocol: Encrypted Extensions
    TLSv1.3 Record Layer: Handshake Protocol: Finished
Transport Layer Security
    TLSv1.3 Record Layer: Alert (Level: Fatal, Description: Unsupported Extension)
```

</details>

**Comment by zphixon** [comment]
Still seeing this behavior with SSL_OP_NO_SESSION_RESUMPTION_ON_RENEGOTIATION set

**Comment by zphixon** [comment]
Sorry to keep bumping. I've also set the cache mode to SSL_SESS_CACHE_OFF.

**Comment by nhorman** [comment]
I assume you (or your organization have authored both the client and the server in this scenario)?  And that openssl is the SSL library used in both?

SSL_CTX_set_psk_use_session_callback doesn't do (or rather isn't meant to do anything to construct extensions in the ClientHello or ServerHello messages).  Rather that sets the callback used to handle session creation when an early extension

The creation of the psk_key_exchange_modes extension is handled in tls_construct_ctos_psk_kex_modes.  likewise the creation of the pre_shared_key extension is handled by a call to tls_construct_ctos_psk.

OpenSSL adds these in tls_construct_extension(), because the context of building a ClientHello message match the defintions for which the extensions are tagged in the ext_defs array:
```
{   
        TLSEXT_TYPE_psk_kex_modes,
        SSL_EXT_CLIENT_HELLO | SSL_EXT_TLS_IMPLEMENTATION_ONLY
        | SSL_EXT_TLS1_3_ONLY,
        init_psk_kex_modes, tls_parse_ctos_psk_kex_modes, NULL, NULL,
        tls_construct_ctos_psk_kex_modes, NULL
    },
```

Note that SSL_EXT_SERVER_HELLO isn't listed there, so its not included in the ServerHello message

As for why the pre_shared_key extension is offered in the ServerHello despite not being solicited, it seems that it is unilaterally included, based on its context defintion:
```
{
        /* Required by the TLSv1.3 spec to always be the last extension */
        TLSEXT_TYPE_psk,
        SSL_EXT_CLIENT_HELLO | SSL_EXT_TLS1_3_SERVER_HELLO
        | SSL_EXT_TLS_IMPLEMENTATION_ONLY | SSL_EXT_TLS1_3_ONLY,
        NULL, tls_parse_ctos_psk, tls_parse_stoc_psk, tls_construct_stoc_psk,
        tls_construct_ctos_psk, final_psk
    }
```

Note its inclusion is limited to TLS1.3.  My guess is in openssl 1.1.1, you may have been using TLS1.2, for which the extension is excluded

Hope that helps

**Reply by zphixon** [reply]
This is helpful, thank you :) Your assumptions are correct, our application is a consumer of OpenSSL. 

I noticed that in `tls_construct_stoc_psk` the PSK extension is included if `s->hit` is true, which seems to imply the server will reply with the PSK extension if the session was cached in some way. I'm guessing that the session state got out of sync somehow, like the client side believed it was starting a new session during the second handshake, but the server thought it was part of the same session.

I tried `SSL_free` and `SSL_new` on the server side between the first and second handshakes to guarantee a new session. The server no longer sends the PSK extension in the second handshake, which seems to support my analysis? My understanding of the library internals and TLS 1.3 itself is limited unfortunately.

We are using TLS 1.3 in both scenarios.

**Reply by nhorman** [reply]
yes, hit identifies if a previous session is being resumed, so if you resume a session the "psk" extension will not be sent.

**Reply by zphixon** [reply]
> if you resume a session the "psk" extension will not be sent

Am I reading this incorrectly? If the session is *not* resumed, the extension will not be sent.

```c
EXT_RETURN tls_construct_stoc_psk(SSL *s, WPACKET *pkt, unsigned int context,
                                  X509 *x, size_t chainidx)
{
    if (!s->hit)
        return EXT_RETURN_NOT_SENT;
```

**Reply by mattcaswell** [reply]
> Am I reading this incorrectly? If the session is not resumed, the extension will not be sent.

You are reading it correctly. The psk extension will only be sent in the ServerHello if TLSv1.3 has been negotiated and the server believes it is resuming (which implies that the client must have sent the psk extension in the ClientHello). So your wireshark trace is quite perplexing.

**Reply by zphixon** [reply]
Ok, this was more annoying and complicated on our end than I first thought. This should be what is happening in the failed case. Each letter corresponds to a different `SSL*`.

```
Component 1 | Component 2 | Component 3
------------+-------------+------------
Connect A   | Accept E    |
Shutdown A  | Shutdown E  |
Connect B   |             | Accept F
Connect C   | Accept G    |
Shutdown C  | Shutdown G  |
Connect D   |             | Accept H
~~~~~~~ Connection interruption ~~~~~~~
Connect I   | Accept K    |
Shutdown I  | Shutdown K  |
Connect I   |             | Accept F
Connect J   | Accept L    |
Shutdown J  | Shutdown L  |
Connect J   |             | Accept H
```

Seems random whether the I-F or J-H handshake fails (as in, whether the ServerHello during the accept on F or the accept on H sends an unsolicited PSK extension). I noticed that we don't shutdown either F or H, perhaps that would have an impact on caching? Although I might expect the different client-side `SSL*`s would not be interpreted as belonging to the same TLS session.

**Comment by markus-heintel** [comment]
For me this issue reproduces with OpenSSL 3.5.0 (see attached pcap; filter for "eap or tls" to get the relevant packets).
1. First full handshake (client and server authentication) results in a NewSessionTicket offered by the server.
2. Client does not use the pre_shared_key extension in its ClientHello (this is done on purpose to force a full handshake)
3. Server includes a pre_shared_key extension in the ServerHello indicating it selected the first identity offered by the client (which was never provided with the ClientHello)
4. Client fails with "unsolicited extension", which is the correct behavior in this situation.
From what I understand the ServerHello should never include a pre_shared_key unless offered by the ClientHello (regardless of any server configuration).
Are there any updates on this issue? Let me know if further information is required.
[unsolicited.zip](https://github.com/user-attachments/files/20038342/unsolicited.zip)

One thing that comes to my mind is, that I'm using SSL_clear() to recycle the SSL object used by the server (instead of SSL_free()/SSL_new()). I'm using SSL_clear() on purpose to avoid superfluous memory management operations and I want to keep the registered BIO methods. Is it possible that SSL_clear() does not reset some information? I still think the presence(or not-presence) of pre_shared_key extension in the ClientHello should take precedence.

Note: I put a question into Q&A referring to the inverse scenario (server should reject resumption), which might be related (https://github.com/openssl/openssl/discussions/27563).

**Comment by mattcaswell** [comment, marked as answer]
Finally tracked this down. Fix for this in #27584

**Reply by zphixon** [reply]
Nice!
